Implement pagiation on account import

csillag opened the pull request.

buberdds commented:
UI issues
- horizontal scroll bar in extension
- ![Screenshot 2022-11-03 at 11 56 51](https://user-images.githubusercontent.com/891392/199704532-4bee6aff-de9d-4fe8-a31c-40ae27585cdf.png)
- possible to generate fatal error while fetching Ledger accounts (fetch and click next)
- ![Screenshot 2022-11-03 at 11 58 48](https://user-images.githubusercontent.com/891392/199704551-c392bca1-97df-49f0-9360-4155a65bf71e.png)
- UI is flickering while loading balances

buberdds commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -127,6 +158,8 @@ export function* sign<T>(signer: LedgerSigner, tw: oasis.consensus.TransactionWr
 export function* importAccountsSaga() {
-  yield* takeEvery(importAccountsActions.enumerateAccountsFromLedger, enumerateAccountsFromLedger)
+  yield* takeEvery(importAccountsActions.enumerateFirstAccountsFromLedger, enumerateSomeAccountsFromLedger)
+  yield* takeEvery(importAccountsActions.enumerateMoreAccountsFromLedger, enumerateSomeAccountsFromLedger)
```
Do we need two separate actions for Ledger? Maybe we don't need `accounts = []`  in `enumerateAccountsFromLedger`  if that was a case.

csillag replied:
There are two entry points for Ledger-related actions:
 - When we start the process
 - When we move to a new page.

For those two situations, I think we do need two separate actions.

However, I can rearrange the functionality so that only one of the two actions is connected to a saga.
 - The first action (which is only triggered when we start the process) is only resetting everything
 - The second action gets the new accounts. (This is triggered both from the first action and when going to next page.)

I will do that now.

csillag replied:
Done.

csillag replied:
So now there is only one saga for the ledger import, and only one trigger action for that saga. So what happens on the first page is exactly the same as what happens on the second page.

However, the drawback is that every time we want to start the ledger account import, we need to dispatch two actions side by side:
```
dispatch(importAccountsActions.resetLedgerAccountImport())
dispatch(importAccountsActions.enumerateAccountsFromLedger())
```
I think this is error a bit prone; it would probably be better to go back to having two actions (one for starting the import process, and another for continuing on the next page)

csillag replied:
I have moved back to having two different actions for the two different entry points.

buberdds commented on src/app/state/importaccounts/index.ts:
```diff
@@ -15,23 +23,37 @@ const slice = createSlice({
     },
+    updateAccount(state, action: PayloadAction<ImportAccountChange>) {
+      const change = action.payload
+      Object.assign(state.accounts[change.index], change)
```
we care only about balance 
`state.accounts[action.payload.index].balance = action.payload.balance`

csillag replied:
Yes, thanks, already fixed that.

buberdds commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -71,33 +77,58 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
+      const balance = yield* call(getBalance, hex2uint(account.publicKey))
+      yield* put(
+        importAccountsActions.updateAccount({
+          index: account.index,
```
do we really need account.index ? Data structure looks like 
0: { index: 0 ...}
1: { index: 1 ...} 
in UI you can extract index from derivation path if you need it in selection modal

csillag replied:
.... right. Simplified that.

csillag commented on src/app/components/Toolbar/Features/AccountSelector/index.tsx:
```diff
@@ -70,6 +71,11 @@ export const Account = memo((props: AccountProps) => {
       )}
+      <Box alignSelf="center" pad={{ left: 'small', right: 'small' }} style={{ minWidth: '2.5em' }}>
+        <Text weight="bold" style={{ width: '2em' }}>
+          {props.index}
```
I don't know, the ClickUp task had this image:
![image](https://user-images.githubusercontent.com/2093792/199723330-68d8527e-0374-4f5b-8acf-21ae60fa39d1.png)

buberdds replied:
I was referring to "Switch to another account" view. Maybe while importing accounts it makes sense, but we have derivation paths.

csillag commented:
> possible to generate fatal error while fetching Ledger accounts (fetch and click next)

should be fixed

> fatal error when opening account while balance is loading

should be fixed

> UI is flickering while loading balances

Can you show me what do you mean by flickering?

csillag commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -8,28 +8,32 @@ import { ImportAccountsStepFormatter } from 'app/components/ImportAccountsStepFo
 import { walletActions } from 'app/state/wallet'
-import { Box, Button, Spinner, Text } from 'grommet'
+import { Anchor, Box, Button, ResponsiveContext, Spinner, Text } from 'grommet'
+import { numberOfPages } from '../../../../state/importaccounts/saga'
```
OK, I fixed that. (btw the `../../state` was also present in other locations in the code; I fixed that, too.)

csillag replied:
I'm fine with either way. (My IDE adds relative paths by default, so that's a bit easier for me.)

csillag commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -64,25 +71,39 @@ export function ImportAccountsSelectionModal(props: ImportAccountsSelectionModal
+
+  const onNext = () => {
+    dispatch(importAccountsActions.setPage(pageNum + 1))
+    if (props.type === 'ledger') {
```
Ah OK. I fixed that. (Along with several other locations of the code when the literal was present.)

csillag commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -59,6 +64,7 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
     }
     yield* setStep(ImportAccountsStep.Done)
     yield* put(importAccountsActions.accountsListed(wallets))
+    yield* put(importAccountsActions.setPage(0))
```
The reason for this is that the saga that will fetch the balances is triggered by the `setPage` action.

csillag replied:
... but we can just as well trigger the balance loading directly, and then there is no need to call the `setPage` action.

csillag commented on src/app/components/Toolbar/Features/AccountSelector/index.tsx:
```diff
@@ -78,9 +84,7 @@ export const Account = memo((props: AccountProps) => {
-          <Box>
-            <AmountFormatter amount={props.balance} />
-          </Box>
+          <Box>{props.balance !== undefined ? <AmountFormatter amount={props.balance} /> : <Spinner />}</Box>
```
Right. Fixed. (And we don't even need the `!`, just simpliy `balance`.)

csillag commented on src/app/state/importaccounts/index.ts:
```diff
@@ -3,7 +3,15 @@ import { ErrorPayload } from 'types/errors'
+  accountsSelectionPageNumber: 0,
+}
+
+interface ImportAccountChange extends Partial<ImportAccountsListAccount> {
```
The original idea was that in theory, we can change any part of the account... but yes, since the only usage is to inject the balance, we can simplify it a lot more.

csillag commented on src/app/state/importaccounts/saga.test.ts:
```diff
@@ -21,59 +21,68 @@ describe('importAccounts Sagas', () => {
           [matchers.call.fn(getBalance), {}],
         ])
-        .dispatch(importAccountsActions.enumerateAccountsFromLedger())
+        .dispatch(importAccountsActions.enumerateFirstAccountsFromLedger())
```
Updated.

csillag commented on src/locales/en/translation.json:
```diff
@@ -209,7 +209,10 @@
     "importAccounts": {
+      "next": "Next",
       "openWallets": "Open",
+      "openXWallets": "Open {{ number }} accounts",
```
keys changed.

csillag commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -95,10 +116,30 @@ export function ImportAccountsSelectionModal(props: ImportAccountsSelectionModal
+          {size === 'small' ? (
+            <Button onClick={onPrev} label={'<'} disabled={!canGoPrev} />
+          ) : (
+            <Anchor onClick={onPrev} disabled={!canGoPrev}>
```
I agree that using a button would be both technically easier and more semantically correct, but doesn't it look better as a link? (Again, the screenshot specified as a reference sketch showed a link.)

csillag replied:
OK, for now I made it a button. We can easily re-style it later if needed.

csillag commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/__tests__/__snapshots__/index.test.tsx.snap:
```diff
@@ -372,19 +386,46 @@ exports[`<ImportAccountsSelectionModal  /> should match snapshot 1`] = `
+                  disabled=""
+                  type="button"
+                >
+                  &lt;
```
`a11yTitle` added.

buberdds commented:
> Can you show me what do you mean by flickering?
That's a minor issue, but height of an account row is changing and it looks like account list is "jumping" or flickering while going through different states. on init row height is X px, than with a loader it's Y px and with balance loaded height is different again. Maybe we need some fixed height on a wrapper element that holds spinner and label.

csillag commented:
> horizontal scroll bar in extension

Should be fixed, too

csillag commented:
Rebased on top of the Ledger permission fix.

> UI is flickering while loading balances

Fixed

csillag commented:
Further possible improvements:
 * Add new test cases that actually test the paging behavior
 * Make it so that the balance checks are executed in parallel with ledger account generation

buberdds commented:
Error occur
Open modal -> click next -> click on modal overlay (dark background)
![Screenshot 2022-11-09 at 11 49 05](https://user-images.githubusercontent.com/891392/200810927-849775b9-01f7-46c3-a372-30670c6c483c.png)

Open modal -> close modal immediately by clicking modal overlay
![Screenshot 2022-11-09 at 11 50 32](https://user-images.githubusercontent.com/891392/200811140-543cfd39-5c33-4267-aa0b-1ff4b02837cb.png)

buberdds commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -71,33 +78,64 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
     yield* setStep(ImportAccountsStep.LoadingAccounts)
-    const accounts = yield* call(Ledger.enumerateAccounts, transport)
+    for (let index = start; index < start + accountsPerPage; index++) {
+      const a = (yield* call(Ledger.enumerateAccounts, transport, 1, index))[0]
```
let's use meaningful var names

csillag replied:
Originally I used account, but then I wanted to minimize my changes (compared to the original code) and when back to "a". Going back to "account" now.

buberdds commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -109,7 +147,11 @@ function* enumerateAccountsFromLedger() {
+      try {
+        yield* call([transport, transport.close])
+      } catch (error) {
+        console.log('Error while closing Ledger USB interface:', error)
```
Shouldn't we throw  WalletError?

csillag replied:
No, I think not, because this happens when we are already trying to close things. Most often this is caused by some other event "closing" the connection before we do, like unplugging the USB. But by this time, we have already handled the error in the above code, this is just some cleanup.

csillag commented on src/app/components/Toolbar/Features/AccountSelector/index.tsx:
```diff
@@ -50,6 +51,10 @@ export const Account = memo((props: AccountProps) => {
 
+  const { details } = props
+
+  const index = details ? details.split('/').reverse()[0] : '?'
```
Nice. Thanks.

csillag commented:
The latest change (changing the number of accounts per page) has introduced a cypress test fail. I'll fix that later.

lukaw3d commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -71,33 +78,64 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
 function* enumerateAccountsFromLedger() {
+  const existingAccounts = yield* select(selectImportAccountsFullList)
+  const pageNumber = yield* select(selectImportAccountsPageNumber)
+  if (existingAccounts.length >= (pageNumber + 1) * accountsPerPage) return
```
This seems odd. Add a comment

csillag replied:
OK, added comment

lukaw3d commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -71,33 +78,64 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
 
     yield* setStep(ImportAccountsStep.LoadingAccounts)
-    const accounts = yield* call(Ledger.enumerateAccounts, transport)
+    for (let index = start; index < start + accountsPerPage; index++) {
+      const account = (yield* call(Ledger.enumerateAccounts, transport, 1, index))[0]
```
This is odd. Refactor enumerateAccounts?

And add a ledger-specific accountsPerPage

csillag replied:
> Refactor enumerateAccounts?

Refactored.

> And add a ledger-specific accountsPerPage

Why should it be ledger-specific? (Ledger is not the only limiting factor; we are constrained to 4 acc per page because of the extension sizing issues anyway.) Do we want to introduce different rules for all small screen / big screen / mnemonic / ledger combinations?

lukaw3d replied:
I thought we might be increasing accountsPerPage for mnemonic at some point, and forget about ledger performance issues.

With new decrease, let's add a comment that accountsPerPage needs to be small enough for extension layout and small enough for slow ledgers

csillag replied:
OK, comment added.

lukaw3d commented on src/app/state/importaccounts/saga.test.ts:
```diff
@@ -84,7 +95,20 @@ describe('importAccounts Sagas', () => {
+          path: [44, 474, i],
+          privateKey: '00',
+          publicKey: '00',
+          selected: !i,
```
i === 0

csillag replied:
ok

lukaw3d commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -78,29 +122,43 @@ export function ImportAccountsSelectionModal(props: ImportAccountsSelectionModal
+            onClick={onPrev}
+          />
+          <Box>
+            Page {importAccounts.accountsSelectionPageNumber + 1} / {numberOfAccountPages}
```
Translation

lukaw3d commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -44,32 +49,71 @@ function ImportAccountsSelector({ accounts }: ImportAccountsSelectorSelectorProp
+    }
+  }
+
+  let selectionLabel: string
+  switch (selectedAccounts.length) {
+    case 0:
+      selectionLabel = t('openWallet.importAccounts.noAccountSelected', 'No account selected')
+      break
+    case 1:
+      selectionLabel = t('openWallet.importAccounts.oneAccountSelected', 'One account selected')
+      break
+    default:
+      selectionLabel = t(
+        'openWallet.importAccounts.multipleAccountsSelected',
+        '{{ number }} accounts selected',
+        { number: selectedAccounts.length },
+      )
+  }
```
This looks like reimplementation of plurals https://www.i18next.com/translation-function/plurals

lukaw3d replied:
We don't have any existing plurals, and i18next-scanner doesn't seem to support it correctly via
```tsx
t('aa', {
  count: count,
  defaultValue_0: 'zero1',
  defaultValue_zero: 'zero2',
  defaultValue_1: 'one1',
  defaultValue_one: 'one2',
  defaultValue: 'plural1 {{count}}',
  defaultValue_plural: 'plural2 {{count}}',
  defaultValue_other: 'plural3 {{count}}',
})
```

I came up with this workaround:
```tsx
  // Workaround for i18next-scanner to pickup plurals correctly, because it is missing
  // defaultValue_zero, defaultValue_one, defaultValue_other here:
  // https://github.com/i18next/i18next-scanner/blob/4687b6a/src/parser.js#L502-L503
  t('aa_zero', 'zero')
  t('aa_one', 'one')

  return <div>{t('aa', 'count {{count}}', { count: count })}</div>
```

csillag replied:
OK, I have implemented it using this workaround.

lukaw3d commented on src/app/components/Toolbar/Features/AccountSelector/index.tsx:
```diff
@@ -50,6 +51,10 @@ export const Account = memo((props: AccountProps) => {
     [WalletType.PrivateKey]: t('toolbar.wallets.type.privateKey', 'Private key'),
   }
 
+  const { details } = props
```
Slight preference against extracting `details` into a variable (we still use `({props.details})` below, and we aren't extracting other props)

csillag replied:
OK, not extracting in any more. (In this function.)

lukaw3d commented on src/app/components/Toolbar/Features/AccountSelector/index.tsx:
```diff
@@ -50,6 +51,10 @@ export const Account = memo((props: AccountProps) => {
 
+  const { details } = props
+
+  const index = details ? details.split('/').at(-1) : '?'
```
I'm not a fan of the name `index` outside of loop iterators. accountNumber?

csillag replied:
OK, renamed.

lukaw3d commented on src/app/state/importaccounts/index.ts:
```diff
@@ -3,41 +3,65 @@ import { ErrorPayload } from 'types/errors'
   initialState,
   reducers: {
-    clear(state, action: PayloadAction<void>) {
+    clear(state, _action: PayloadAction<void>) {
```
You can remove the _action parameters :D

csillag replied:
I have removed most of them, except this one: 
```
enumerateAccountsFromMnemonic(state, _action: PayloadAction<string>) 
```
... since keeping this is necessary to properly define the type of this action. (The string payload is used by the saga.)

lukaw3d commented on src/app/state/importaccounts/saga.ts:
```diff
@@ -71,33 +78,64 @@ function* enumerateAccountsFromMnemonic(action: PayloadAction<string>) {
+function* ensureAllBalancesArePresentOnCurrentPage() {
+  const existingAccounts = yield* select(selectImportAccountsOnCurrentPage)
+  const accountsWithNoBalance = existingAccounts.filter(a => !a.balance)
+  for (let i = 0; i < accountsWithNoBalance.length; i++) {
+    yield* fork(fetchBalanceForAccount, accountsWithNoBalance[i])
+  }
```
Does this intentionally not wait? I'd expect `yield* all([ ... ])`

csillag replied:
Re-implemented using `yield* all()`. (The flow was the same both before and after the change.)

lukaw3d commented on src/app/pages/OpenWalletPage/Features/ImportAccountsSelectionModal/index.tsx:
```diff
@@ -44,32 +49,71 @@ function ImportAccountsSelector({ accounts }: ImportAccountsSelectorSelectorProp
+  const cancelDisabled = isBusyImporting && !error
+  const confirmDisabled = isBusyImporting || selectedAccounts.length === 0
+
+  const canGoPrev = !isBusyImporting && !!pageNum && !error
+
+  const onPrev = () => {
+    dispatch(importAccountsActions.setPage(pageNum - 1))
+  }
+
+  const canGoNext = !isBusyImporting && pageNum !== numberOfAccountPages - 1 && !error
```
```ts
const canGoPrev = pageNum > 0 && !isBusyImporting && !error
const canGoNext = pageNum < numberOfPages && !isBusyImporting && !error
```
would be more readable

csillag replied:
OK, updated. (Except that we need `< numberOfPages - 1`.)

lukaw3d commented:
> @buberdds Error occur Open modal -> click next -> click on modal overlay (dark background)
> 
> Open modal -> close modal immediately by clicking modal overlay

We should probably just disable clicking outside modal and <kbd>Esc</kbd>.
That would fix: https://github.com/oasisprotocol/oasis-wallet-web/issues/869

csillag commented:
Changing the number of accounts per page (from 5 to 4) has been separated in to a separate commit: https://github.com/oasisprotocol/oasis-wallet-web/pull/1124/commits/2d9c918df596d04c94c30d66f9ff67e78b658e3b

csillag commented:
> We should probably just disable clicking outside modal and Esc. That would fix: #869

We can do that in a separate PR, right?

lukaw3d commented:
> > We should probably just disable clicking outside modal and Esc. That would fix: #869
> 
> We can do that in a separate PR, right?

Yeah, just make sure edge cases from this PR don't remain

lukaw3d reviewed (commented):
I've split the commits in a more sensible manner:
https://github.com/oasisprotocol/oasis-wallet-web/compare/lw/split-pagination

diff of changes: https://github.com/oasisprotocol/oasis-wallet-web/compare/fda9835..cb3f1e8

lukaw3d commented:
Updated previews

| Mnemonic | Ledger | Extension layout |
| --- | --- | --- |
| ![mnemonic](https://user-images.githubusercontent.com/3758846/200977688-75908776-c333-4733-af08-0718e9cbb7f8.gif)  | ![ledger](https://user-images.githubusercontent.com/3758846/200977684-eccc2467-64c8-4620-bed9-86faa92a37d6.gif)  | ![image](https://user-images.githubusercontent.com/3758846/200977672-b0a5de53-96a0-43d5-9ddb-638e8e1ce8e3.png) |

buberdds reviewed (approved):
issues reported by me are fixed

csillag commented:
> I've split the commits in a more sensible manner:

Indeed, it does make more sense that way. (I have replaced the commits of this branch with that one.)

csillag commented:
> Could we hardcode the accounts window size while populating it with Ledger? Moving clickable buttons are PITA.

I think I have fixed that. (But I can't test with a Ledger right now.)
